src/Glpi/Controller/GenericFormController.php

```diff
@@ -0,0 +1,187 @@
+        $object->check($id, self::ACTIONS_AND_CHECKS[$form_action]['permission'] ?? READ, $post_data);
+
+        // Special case for GET
+        if ($form_action === 'get' && $request->getMethod() === 'GET') {
+            return $this->displayForm($object, $request, $isTemplateForm);
+        }
+
+        // POST action execution
+        $action_result = match ($form_action) {
+            'add' => $object->add($post_data),
+            'delete' => $object->delete($post_data),
+            'restore' => $object->restore($post_data),
+            'purge' => $object->delete($post_data, 1),
+            'update' => $object->update($post_data),
+            'unglobalize' => $object->unglobalize(),
+            default => throw new \RuntimeException(\sprintf("Unsupported object action \"%s\".", $form_action)),
+        };
+
+        if ($action_result) {
+            Event::log(
+                $action_result,
+                \strtolower(\basename($class)),
+                $object::getLogLevel(),
+                $object::getLogServiceName(),
+                sprintf(__('%1$s executes the "%2$s" action on the item %3$s'), $_SESSION["glpiname"], $form_action, $post_data["name"])
+            );
+        }
+
+        // Specific case for "add"
+        if ($action_result && $form_action === 'add' && $_SESSION['glpibackcreated']) {
+            return new RedirectResponse($object->getLinkURL());
+        }
+
+        $post_action = $object::getPostFormAction($form_action) ?? 'list';
+        if (!\in_array($post_action, ['back', 'list', 'form'])) {
+            $post_action = 'list';
+        }
+
+        return match ($post_action) {
+            'back' => new RedirectResponse(Html::getBackUrl()),
+            'form' => new RedirectResponse($object::getFormURLWithID($id)),
+            'list' => new RedirectResponse($object->getRedirectToListUrl()),
+        };
+    }
+
+    /**
+     * @param class-string<CommonDBTM> $class
+     */
+    private function getCurrentAllowedAction(Request $request, string $class): ?string
+    {
+        if ($request->getMethod() === 'POST') {
+            foreach (\array_keys(self::ACTIONS_AND_CHECKS) as $action) {
+                if (
+                    $request->request->has($action)
+                    && \method_exists($class, $action)
+                ) {
+                    return $action;
+                }
+            }
+        }
+
+        return $request->getMethod() === 'GET' ? 'get' : null;
+    }
+
+    public function displayForm(CommonDBTM $object, Request $request, bool $isTemplateForm): Response
+    {
+        $form_options = $object->getFormOptionsFromUrl($request->query->all());
+        $form_options['formoptions'] = 'data-track-changes=true';
+        if ($isTemplateForm) {
+            $form_options['withtemplate'] = $request->query->get('withtemplate', '');
+        }
+
+        return $this->render('pages/generic_form.html.twig', [
+            'id' => $request->query->get('id', 0),
```

Comment:
`-1` will be safer, in case this controller is used for `Entity` in the future.

```suggestion
            'id' => $request->query->get('id', -1),
```